### PR TITLE
OP-781 drop LAB_EXAM_DATE field

### PR DESCRIPTION
**Pull request by dbmalkovsky**
See OP-781

All CORE unit tests passed locally.

There is a corresponding GUI PR that fixes the JasperReports given the `LAB_EXAM_DATE` field is removed.

### CODE REVIEW

File: src/main/java/org/isf/lab/manager/LabManager.java [hunk 1]
```diff
@@ -79,8 +79,8 @@ protected void setPatientConsistency(Laboratory laboratory) {
 	 */
 	protected void validateLaboratory(Laboratory laboratory) throws OHDataValidationException {
 		List<OHExceptionMessage> errors = new ArrayList<>();
-		if (laboratory.getExamDate() == null) {
-			laboratory.setExamDate(LocalDate.now());
+		if (laboratory.getCreatedDate() == null) {
+			laboratory.setCreatedDate(TimeTools.getNow());
```
Comment (mwithi) [inline src/main/java/org/isf/lab/manager/LabManager.java:83]: This should be managed by JPA automatically

File: src/main/java/org/isf/lab/model/Laboratory.java [hunk 2]
```diff
@@ -270,7 +249,7 @@ public int hashCode() {
 	@Override
 	public String toString() {
 		return "-------------------------------------------\nLaboratory{" + "code=" + code + ", material=" + material
-				+ ", exam=" + exam + ", registrationDate=" + labDate + ", examDate=" + examDate + ", result="
+				+ ", exam=" + exam + ", registrationDate=" + labDate + ", examDate=" + createdDate + ", result="
```
Comment (mwithi) [inline src/main/java/org/isf/lab/model/Laboratory.java:252]: this should be `labDate`, simply

File: src/test/java/org/isf/lab/test/TestLaboratory.java [hunk 3]
```diff
@@ -39,7 +38,7 @@ public class TestLaboratory {
 
 	private String material = "TestMaterial";
 	private LocalDateTime labDate = TimeTools.getNow();
-	private LocalDate examDate = labDate.toLocalDate();
+	private LocalDateTime examDate = labDate;
```
Comment (mwithi) [inline src/test/java/org/isf/lab/test/TestLaboratory.java:41]: is this `examDate` still used?

File: src/test/java/org/isf/lab/test/TestLaboratory.java [hunk 4]
```diff
@@ -94,7 +91,9 @@ public void check(Laboratory laboratory) {
 			assertThat(laboratory.getPatName()).isEqualTo(patName);
 		}
 		assertThat(laboratory.getDate()).isCloseTo(labDate, within(1, ChronoUnit.SECONDS));
-		assertThat(laboratory.getExamDate()).isEqualTo(examDate);
+		// because CreatedDate is set by the auditable class when the datbase is updated the value will differ from
+		// the static examDate value set above.
+		assertThat(laboratory.getCreatedDate()).isCloseTo(examDate, within(25, ChronoUnit.SECONDS));
```
Comment (mwithi) [inline src/test/java/org/isf/lab/test/TestLaboratory.java:96]: I think auditable fields can be ignored in tests

File: src/main/java/org/isf/lab/manager/LabManager.java (hunk 1 above)
Comment (mwithi) [reply src/main/java/org/isf/lab/manager/LabManager.java:83]: where `setCreatedDate` is defined?

Comment (dbmalkovsky) [reply src/main/java/org/isf/lab/manager/LabManager.java:83]: Use you are correct it can be removed.

`setCreateDate` is found in `org.isf.utils.db.Auditable`

File: src/main/java/org/isf/lab/model/Laboratory.java (hunk 2 above)
Comment (dbmalkovsky) [reply src/main/java/org/isf/lab/model/Laboratory.java:252]: Fixed.

File: src/test/java/org/isf/lab/test/TestLaboratory.java (hunk 4 above)
Comment (dbmalkovsky) [reply src/test/java/org/isf/lab/test/TestLaboratory.java:96]: Agreed; good catch.

File: src/main/java/org/isf/lab/model/Laboratory.java (hunk 2 above)
Comment (mwithi) [reply src/main/java/org/isf/lab/model/Laboratory.java]: Thanks!

File: src/test/java/org/isf/lab/test/TestLaboratory.java (hunk 3 above)
Comment (mwithi) [reply src/test/java/org/isf/lab/test/TestLaboratory.java]: Thanks!

File: sql/step_78_adjust_lab_datetime.sql [hunk 5]
```diff
@@ -0,0 +1,5 @@
+UPDATE OH_LABORATORY SET LAB_CREATED_DATE = LAB_DATE;
+UPDATE OH_LABORATORY SET LAB_DATE = CAST(CONCAT(LAB_EXAM_DATE, ' ', TIME(LAB_DATE)) AS DATETIME);
+
+ALTER TABLE OH_LABORATORY
+DROP COLUMN LAB_EXAM_DATE;
```
Comment (mwithi) [inline sql/step_78_adjust_lab_datetime.sql:5]: I would put it in a singole line
and I would rename the file in `step_78_drop_lab_exam_date.sql` (like step_71)

**Comment by mwithi** [comment]
IDK if it is this the expected result, but now labDate is second-less
image
which will always differ from created_date
image

This might be true for all new `GoodTime*Chooser` though...

File: sql/step_78_adjust_lab_datetime.sql (hunk 5 above)
Comment (dbmalkovsky) [reply sql/step_78_adjust_lab_datetime.sql:5]: done; BTW, I just used the name that was there in the extra directory and changed XX to 78

Comment (mwithi) [reply sql/step_78_adjust_lab_datetime.sql]: ok, thanks!

**Comment by dbmalkovsky** [comment]
We are storing the seconds:  20:53:49 is 10PM, 53 minutes and 49 seconds.

**Comment by mwithi** [comment]
> We are storing the seconds: 20:53:49 is 8 PM, 53 minutes and 49 seconds.

yes, but not in LAB_DATE

**Comment by dbmalkovsky** [comment]
I just created a new lab and it looks like there are seconds:
image

**Comment by mwithi** [comment]
> I just created a new lab and it looks like there are seconds: ![image](https://user-

let me try again...

**Comment by mwithi** [comment]
> > I just created a new lab and it looks like there are seconds: image -> again JDK problem?
2. subsequent exams will get always the same time -> I think because of RememberDates, but it should keep only the date part

**Comment by mwithi** [comment]
by debuggin it seems the value is taken directly like that 🤔

image

**Comment by mwithi** [comment]
The same happens in OPD (and I suppose everywhere)

image

File: src/main/java/org/isf/lab/manager/LabManager.java (hunk 1 above)
Comment (mwithi) [reply src/main/java/org/isf/lab/manager/LabManager.java:83]: is this removed or not yet?

File: src/test/java/org/isf/lab/test/Tests.java [hunk 6]
```diff
@@ -696,7 +696,7 @@ public void testMgrValidationMissingExamDate() throws Exception {
 
 		ArrayList<String> labRow = new ArrayList<>();
 		Laboratory laboratory = testLaboratory.setup(exam, patient, false);
-		laboratory.setExamDate(null);
+		laboratory.setCreatedDate(null);
```
Comment (mwithi) [inline src/test/java/org/isf/lab/test/Tests.java:699]: This also I think can be removed

File: src/main/java/org/isf/lab/model/Laboratory.java (hunk 2 above)
Comment (mwithi) [reply src/main/java/org/isf/lab/model/Laboratory.java]: sorry, spotted now... `registrationDate=" + createdDate + ", examDate=" + labDate`
